File: tests/js/test_bindings.mjs
// Tests for the v0.1 Node bindings.
//
// Mirrors tests/python/test_bindings.py. Run with `npm test`.

import { test } from 'node:test'
import assert from 'node:assert/strict'
import fs from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'

import { Shell, ShellError, NotFoundError, FileTooLargeError } from '../../index.js'

async function makeHostDir() {
  return fs.mkdtemp(path.join(os.tmpdir(), 'strands-shell-bindings-test-'))
}

async function makeShell(hostDir) {
  return Shell.create({
    binds: [{ source: hostDir, destination: '/work', mode: 'direct' }],
    timeout: 10.0,
  })
}

const enc = (s) => new TextEncoder().encode(s)
const dec = (b) => new TextDecoder().decode(b)

// ---------------------------------------------------------------------------
// readFile
// ---------------------------------------------------------------------------

test('readFile returns Uint8Array', async () => {
  const hostDir = await makeHostDir()
  try {
    await fs.writeFile(path.join(hostDir, 'seed.txt'), 'hello from host')
    const shell = await makeShell(hostDir)
    const data = await shell.readFile('/work/seed.txt')
    assert.ok(data instanceof Uint8Array)
    assert.equal(dec(data), 'hello from host')
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('readFile missing rejects with NotFoundError', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await assert.rejects(
      () => shell.readFile('/work/does-not-exist'),
      (err) => {
        assert.ok(err instanceof NotFoundError, 'not a NotFoundError')
        assert.ok(err instanceof ShellError, 'not a ShellError')
        assert.equal(err.code, 'ENOENT')
        assert.equal(err.path, '/work/does-not-exist')
        return true
      },
    )
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('readFile preserves full byte range (binary roundtrip)', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    const payload = new Uint8Array(256)
    for (let i = 0; i < 256; i++) payload[i] = i
    await shell.writeFile('/work/binary.bin', payload)
    const got = await shell.readFile('/work/binary.bin')
    assert.deepEqual(Array.from(got), Array.from(payload))
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('readFile respects maxFileSize (rejects with FileTooLargeError)', async () => {
  // Seed on the host so the file already exceeds the cap at read time; the
  // read must be bounded by maxFileSize instead of loading it all into memory
  // — important for direct-passthrough mounts to large host files.
  const hostDir = await makeHostDir()
  try {
    await fs.writeFile(path.join(hostDir, 'big.txt'), 'x'.repeat(4096))
    const shell = await Shell.create({
      binds: [{ source: hostDir, destination: '/work', mode: 'direct' }],
      limits: { maxFileSize: 1024 },
    })
    await assert.rejects(
      () => shell.readFile('/work/big.txt'),
      (err) => {
        assert.ok(err instanceof FileTooLargeError, `expected FileTooLargeError, got ${err.name}`)
        assert.equal(err.code, 'EFBIG')
        assert.equal(err.path, '/work/big.txt')
        return true
      },
    )
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('readFile at maxFileSize boundary succeeds', async () => {
  const hostDir = await makeHostDir()
  try {
    await fs.writeFile(path.join(hostDir, 'exact.txt'), 'y'.repeat(1024))
    const shell = await Shell.create({
      binds: [{ source: hostDir, destination: '/work', mode: 'direct' }],
      limits: { maxFileSize: 1024 },
    })
    const got = await shell.readFile('/work/exact.txt')
    assert.equal(got.length, 1024)
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

// ---------------------------------------------------------------------------
// writeFile
// ---------------------------------------------------------------------------

test('writeFile creates parent directories', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await shell.writeFile('/work/deep/dir/binary.bin', enc('data'))
    const got = await shell.readFile('/work/deep/dir/binary.bin')
    assert.equal(dec(got), 'data')
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('writeFile truncates on overwrite', async () => {
  const hostDir = await makeHostDir()
  try {
    await fs.writeFile(path.join(hostDir, 'seed.txt'), 'hello from host')
    const shell = await makeShell(hostDir)
    await shell.writeFile('/work/seed.txt', enc('short'))
    const got = await shell.readFile('/work/seed.txt')
    assert.equal(dec(got), 'short')
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('writeFile accepts empty payload', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await shell.writeFile('/work/empty.txt', new Uint8Array(0))
    const got = await shell.readFile('/work/empty.txt')
    assert.equal(got.length, 0)
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('writeFile at root of bind mount', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    const data = new Uint8Array([0, 1, 2])
    await shell.writeFile('/work/root_level.bin', data)
    const got = await shell.readFile('/work/root_level.bin')
    assert.deepEqual(Array.from(got), [0, 1, 2])
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('writeFile handles 2 MiB payload (exercises stall-detection bound)', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    const big = new Uint8Array(2 * 1024 * 1024)
    for (let i = 0; i < big.length; i++) big[i] = (i * 31) & 0xff
    await shell.writeFile('/work/big.bin', big)
    const got = await shell.readFile('/work/big.bin')
    assert.equal(got.length, big.length)
    // Spot-check a few bytes; full deepEqual is slow on 2 MiB.
    assert.equal(got[0], 0)
    assert.equal(got[1], 31)
    assert.equal(got[big.length - 1], big[big.length - 1])
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('writeFile pure-VFS path (no bind mount)', async () => {
  const shell = await Shell.create({ timeout: 10.0 })
  await shell.writeFile('/tmp/vfs_only.txt', enc('in-memory'))
  const got = await shell.readFile('/tmp/vfs_only.txt')
  assert.equal(dec(got), 'in-memory')
})

test('writeFile size-limit rejects with FileTooLargeError, not a timeout', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await Shell.create({
      binds: [{ source: hostDir, destination: '/work', mode: 'copy' }],
      limits: { maxFileSize: 64 },
    })
    await assert.rejects(
      () => shell.writeFile('/work/too_big.bin', new Uint8Array(1024).fill(0x78)),
      (err) => {
        assert.ok(err instanceof FileTooLargeError, `expected FileTooLargeError, got ${err.name}`)
        assert.equal(err.code, 'EFBIG')
        return true
      },
    )
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

// ---------------------------------------------------------------------------
// removeFile
// ---------------------------------------------------------------------------

test('removeFile removes entry', async () => {
  const hostDir = await makeHostDir()
  try {
    await fs.writeFile(path.join(hostDir, 'seed.txt'), 'hello')
    const shell = await makeShell(hostDir)
    await shell.removeFile('/work/seed.txt')
    const names = (await shell.listFiles('/work')).map(e => e.name)
    assert.ok(!names.includes('seed.txt'))
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('removeFile missing rejects with NotFoundError', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await assert.rejects(
      () => shell.removeFile('/work/does-not-exist'),
      (err) => err instanceof NotFoundError && err.code === 'ENOENT',
    )
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

// ---------------------------------------------------------------------------
// listFiles
// ---------------------------------------------------------------------------

test('listFiles returns structured FileInfo with size and isDir', async () => {
  const hostDir = await makeHostDir()
  try {
    await fs.writeFile(path.join(hostDir, 'seed.txt'), 'short')  // 5 bytes
    await fs.mkdir(path.join(hostDir, 'sub'))
    const shell = await makeShell(hostDir)
    const entries = await shell.listFiles('/work')
    const byName = new Map(entries.map(e => [e.name, e]))
    assert.ok(byName.has('seed.txt'))
    assert.ok(byName.has('sub'))
    assert.equal(byName.get('seed.txt').isDir, false)
    assert.equal(byName.get('sub').isDir, true)
    assert.equal(byName.get('seed.txt').size, 5)
    // Directories don't carry a size; napi-rs surfaces Option::None as
    // the property being absent (undefined), not null.
    assert.equal(byName.get('sub').size, undefined)
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('listFiles on nested directory', async () => {
  const hostDir = await makeHostDir()
  try {
    await fs.mkdir(path.join(hostDir, 'sub'))
    await fs.writeFile(path.join(hostDir, 'sub', 'nested.txt'), 'nested file')
    const shell = await makeShell(hostDir)
    const nested = await shell.listFiles('/work/sub')
    assert.deepEqual(nested.map(e => e.name).sort(), ['nested.txt'])
    assert.equal(nested[0].isDir, false)
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('listFiles missing rejects with NotFoundError', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await assert.rejects(
      () => shell.listFiles('/work/no/such/dir'),
      (err) => err instanceof NotFoundError && err.code === 'ENOENT',
    )
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

// ---------------------------------------------------------------------------
// run() interleaves with native VFS calls; state persists
// ---------------------------------------------------------------------------

test('run interleaves with native VFS calls', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await shell.writeFile('/work/from_node.txt', enc('hi'))
    const out = await shell.run('ls /work')
    assert.equal(out.status, 0)
    assert.match(out.stdout, /from_node\.txt/)
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('cwd persists across run() calls', async () => {
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await shell.run('cd /work')
    const out = await shell.run('pwd')
    assert.equal(out.status, 0)
    assert.match(out.stdout, /\/work/)
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('Buffer is accepted as Uint8Array on writeFile', async () => {
  // Node Buffer extends Uint8Array, so passing a Buffer should work
  // without conversion. This is a key DX guarantee from the doc.
  const hostDir = await makeHostDir()
  try {
    const shell = await makeShell(hostDir)
    await shell.writeFile('/work/buf.txt', Buffer.from('from a Buffer'))
    const got = await shell.readFile('/work/buf.txt')
    assert.equal(Buffer.from(got).toString('utf8'), 'from a Buffer')
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('configFile values are not clobbered by omitted options', async () => {
  // Regression: omitting umask/timeout/limits must NOT overwrite what the
  // TOML set. A value present only in the file must survive.
  const hostDir = await makeHostDir()
  try {
    const cfg = path.join(hostDir, 'shell.toml')
    await fs.writeFile(cfg, 'umask = "077"\n')
    const shell = await Shell.create({ configFile: cfg })
    const out = await shell.run('umask')
    assert.equal(out.stdout.trim(), '0077')
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

test('explicit option overrides configFile', async () => {
  const hostDir = await makeHostDir()
  try {
    const cfg = path.join(hostDir, 'shell.toml')
    await fs.writeFile(cfg, 'umask = "077"\n')
    const shell = await Shell.create({ configFile: cfg, umask: 0o022 })
    const out = await shell.run('umask')
    assert.equal(out.stdout.trim(), '0022')
  } finally {
    await fs.rm(hostDir, { recursive: true, force: true })
  }
})

// ---------------------------------------------------------------------------
// timeout validation (Shell.create rejects non-positive / non-finite)
// ---------------------------------------------------------------------------

test('Shell.create rejects timeout: 0', async () => {
  await assert.rejects(
    () => Shell.create({ timeout: 0 }),
    /timeout must be a positive, finite number/,
  )
})

test('Shell.create rejects negative / non-finite timeout', async () => {
  for (const bad of [-1, NaN, Infinity, -Infinity]) {
    await assert.rejects(
      () => Shell.create({ timeout: bad }),
      /timeout must be a positive, finite number/,
      `timeout: ${bad} should be rejected`,
    )
  }
})

test('Shell.create allows omitted timeout (no limit) and positive values', async () => {
  const noTimeout = await Shell.create({})
  assert.equal((await noTimeout.run('echo ok')).stdout.trim(), 'ok')
  const withTimeout = await Shell.create({ timeout: 5.0 })
  assert.equal((await withTimeout.run('echo ok')).stdout.trim(), 'ok')
})

File: tests/python/test_bindings.py
"""Pytest suite for the v0.1 Python bindings.

Covers the four file-operation methods on `strands_shell.Shell` plus the `FileInfo`
pyclass. Run with:

    .venv/bin/pytest tests/python -v

Each test builds its own `Shell` so failures stay isolated.
"""

import math
import os
import shutil
import tempfile

import pytest

import strands_shell


@pytest.fixture
def host_dir():
    """A throwaway host directory bound to /work in the shell's VFS."""
    path = tempfile.mkdtemp(prefix="strands-shell-bindings-test-")
    try:
        yield path
    finally:
        shutil.rmtree(path, ignore_errors=True)


@pytest.fixture
def shell(host_dir):
    """A shell with `host_dir` bound to /work via passthrough.

    `bind_direct` (rather than copy-mode `bind`) keeps host and VFS in sync,
    which lets tests seed via the host filesystem and observe through the
    shell. This matches how Strands users will normally use Strands Shell.

    Uses the config-driven constructor that is the public Python API, so every
    test that uses this fixture implicitly exercises it.
    """
    return strands_shell.Shell(
        binds=[strands_shell.Bind(host_dir, "/work", mode="direct")],
        timeout=10.0,
    )


def _seed(host_dir, rel_path, content):
    full = os.path.join(host_dir, rel_path)
    os.makedirs(os.path.dirname(full), exist_ok=True) if os.path.dirname(rel_path) else None
    with open(full, "wb") as f:
        f.write(content if isinstance(content, bytes) else content.encode())


# ---------------------------------------------------------------------------
# read_file
# ---------------------------------------------------------------------------

def test_read_file_returns_bytes(host_dir, shell):
    _seed(host_dir, "seed.txt", "hello from host")
    data = shell.read_file("/work/seed.txt")
    assert isinstance(data, bytes)
    assert data == b"hello from host"


def test_read_file_missing_raises_not_found(shell):
    # Typed: strands_shell.FileNotFoundError, which also inherits builtins.FileNotFoundError.
    with pytest.raises(strands_shell.FileNotFoundError) as exc_info:
        shell.read_file("/work/does-not-exist")
    err = exc_info.value
    assert isinstance(err, strands_shell.ShellError)
    assert isinstance(err, FileNotFoundError)  # stdlib builtin
    assert err.path == "/work/does-not-exist"


def test_read_file_preserves_full_byte_range(shell):
    payload = bytes(range(256))
    shell.write_file("/work/binary.bin", payload)
    assert shell.read_file("/work/binary.bin") == payload


def test_read_file_respects_max_file_size(host_dir):
    """A read larger than max_file_size must surface as FileTooLargeError
    rather than loading an unbounded payload into memory — important for
    direct-passthrough mounts to large host files. Seeded on the host so the
    file already exceeds the cap at read time."""
    _seed(host_dir, "big.txt", b"x" * 4096)
    shell = strands_shell.Shell(
        binds=[strands_shell.Bind(host_dir, "/work", mode="direct")],
        limits=strands_shell.Limits(max_file_size=1024),
    )
    with pytest.raises(strands_shell.FileTooLargeError) as exc_info:
        shell.read_file("/work/big.txt")
    assert isinstance(exc_info.value, strands_shell.ShellError)
    assert exc_info.value.path == "/work/big.txt"


def test_read_file_at_max_file_size_boundary_succeeds(host_dir):
    """A read exactly at the cap is allowed."""
    _seed(host_dir, "exact.txt", b"y" * 1024)
    shell = strands_shell.Shell(
        binds=[strands_shell.Bind(host_dir, "/work", mode="direct")],
        limits=strands_shell.Limits(max_file_size=1024),
    )
    assert shell.read_file("/work/exact.txt") == b"y" * 1024


# ---------------------------------------------------------------------------
# write_file
# ---------------------------------------------------------------------------

def test_write_file_creates_parent_directories(shell):
    shell.write_file("/work/deep/dir/binary.bin", b"data")
    assert shell.read_file("/work/deep/dir/binary.bin") == b"data"


def test_write_file_truncates_on_overwrite(host_dir, shell):
    _seed(host_dir, "seed.txt", "hello from host")
    shell.write_file("/work/seed.txt", b"short")
    assert shell.read_file("/work/seed.txt") == b"short"


def test_write_file_accepts_empty_payload(shell):
    shell.write_file("/work/empty.txt", b"")
    assert shell.read_file("/work/empty.txt") == b""


def test_write_file_at_root_of_bind_mount(shell):
    shell.write_file("/work/root_level.bin", b"\x00\x01\x02")
    assert shell.read_file("/work/root_level.bin") == b"\x00\x01\x02"


def test_write_file_handles_large_payload(shell):
    """2 MiB exceeds the 8 KiB drain pipe; exercises the stall-detection bound."""
    big = bytes((i * 31) & 0xff for i in range(2 * 1024 * 1024))
    shell.write_file("/work/big.bin", big)
    assert shell.read_file("/work/big.bin") == big


def test_write_file_pure_vfs_path(shell):
    """Pure-VFS write (no bind mount) goes through the kernel's in-memory drain."""
    shell.write_file("/tmp/vfs_only.txt", b"in-memory")
    assert shell.read_file("/tmp/vfs_only.txt") == b"in-memory"


def test_write_file_size_limit_surfaces_clear_error(host_dir):
    """Writing past max_file_size must error as FileTooLargeError, not look
    like a timeout."""
    shell = strands_shell.Shell(
        binds=[strands_shell.Bind(host_dir, "/work", mode="copy")],
        limits=strands_shell.Limits(max_file_size=64),
    )
    with pytest.raises(strands_shell.FileTooLargeError) as exc_info:
        shell.write_file("/work/too_big.bin", b"x" * 1024)
    assert isinstance(exc_info.value, strands_shell.ShellError)


# ---------------------------------------------------------------------------
# remove_file
# ---------------------------------------------------------------------------

def test_remove_file_removes_entry(host_dir, shell):
    _seed(host_dir, "seed.txt", "hello")
    shell.remove_file("/work/seed.txt")
    names = {e.name for e in shell.list_files("/work")}
    assert "seed.txt" not in names


def test_remove_file_missing_raises_not_found(shell):
    with pytest.raises(strands_shell.FileNotFoundError) as exc_info:
        shell.remove_file("/work/does-not-exist")
    assert isinstance(exc_info.value, strands_shell.ShellError)


# ---------------------------------------------------------------------------
# list_files
# ---------------------------------------------------------------------------

def test_list_files_returns_structured_file_info(host_dir, shell):
    _seed(host_dir, "seed.txt", "short")  # 5 bytes
    os.makedirs(os.path.join(host_dir, "sub"))
    entries = shell.list_files("/work")
    by_name = {e.name: e for e in entries}
    assert "seed.txt" in by_name
    assert "sub" in by_name
    assert by_name["seed.txt"].is_dir is False
    assert by_name["sub"].is_dir is True
    assert by_name["seed.txt"].size == 5
    assert by_name["sub"].size is None  # directories don't carry a size


def test_list_files_on_nested_directory(host_dir, shell):
    os.makedirs(os.path.join(host_dir, "sub"))
    _seed(host_dir, "sub/nested.txt", "nested file")
    nested = shell.list_files("/work/sub")
    assert {e.name for e in nested} == {"nested.txt"}
    assert nested[0].is_dir is False


def test_list_files_missing_raises_not_found(shell):
    with pytest.raises(strands_shell.FileNotFoundError) as exc_info:
        shell.list_files("/work/no/such/dir")
    assert isinstance(exc_info.value, strands_shell.ShellError)


# ---------------------------------------------------------------------------
# Integration with shell.run() and FileInfo repr
# ---------------------------------------------------------------------------

def test_run_interleaves_with_native_vfs_calls(shell):
    shell.write_file("/work/from_python.txt", b"hi")
    out = shell.run("ls /work")
    assert out.status == 0
    assert "from_python.txt" in out.stdout


def test_file_info_repr_is_pythonic(shell):
    shell.write_file("/work/x.txt", b"data")
    entry = next(e for e in shell.list_files("/work") if e.name == "x.txt")
    rep = repr(entry)
    assert rep.startswith("FileInfo(")
    assert "is_dir=False" in rep  # not Some(false)
    assert "size=4" in rep  # not Some(4)


# ---------------------------------------------------------------------------
# config_file merge semantics
# ---------------------------------------------------------------------------

def test_config_file_values_are_not_clobbered_by_defaults(tmp_path):
    """A value set only in config_file must survive — defaulting a constructor
    arg must NOT silently overwrite it. Regression for the merge bug where
    umask/timeout/limits were applied unconditionally."""
    cfg = tmp_path / "shell.toml"
    cfg.write_text('umask = "077"\n')
    shell = strands_shell.Shell(config_file=str(cfg))
    # umask from the file (077) must hold, not the binding's old 0o022 default.
    assert shell.run("umask").stdout.strip() == "0077"


def test_explicit_arg_overrides_config_file(tmp_path):
    """An explicitly passed arg still wins over the config_file value."""
    cfg = tmp_path / "shell.toml"
    cfg.write_text('umask = "077"\n')
    shell = strands_shell.Shell(config_file=str(cfg), umask=0o022)
    assert shell.run("umask").stdout.strip() == "0022"


# --------------------------------------------------------------------------- #
# timeout validation (Shell rejects non-positive / non-finite)
# --------------------------------------------------------------------------- #


def test_zero_timeout_raises_value_error():
    with pytest.raises(ValueError, match="positive, finite"):
        strands_shell.Shell(timeout=0)


@pytest.mark.parametrize("bad", [-1.0, math.nan, math.inf, -math.inf])
def test_negative_or_nonfinite_timeout_raises_value_error(bad):
    with pytest.raises(ValueError, match="positive, finite"):
        strands_shell.Shell(timeout=bad)


def test_omitted_and_positive_timeout_allowed():
    # Omitted timeout => no limit; a positive value => bounded. Both build.
    assert strands_shell.Shell().run("echo ok").stdout.strip() == "ok"
    assert strands_shell.Shell(timeout=5.0).run("echo ok").stdout.strip() == "ok"


# --------------------------------------------------------------------------- #
# Cedar authorization policies
# --------------------------------------------------------------------------- #

_READ_ONLY_POLICY = (
    'permit(principal, action in ['
    'Agent::Action::"fs:read", Agent::Action::"fs:stat", Agent::Action::"fs:list"'
    '], resource);'
)


def test_policy_str_denies_unpermitted_action():
    """An inline read-only policy allows reads but denies writes."""
    shell = strands_shell.Shell(policy=_READ_ONLY_POLICY)
    assert shell.run("ls /").status == 0
    # A write is not permitted by the policy => denied, non-zero exit.
    assert shell.run("echo hi > /home/lash/x.txt").status != 0


def test_policy_file_loads_from_path(tmp_path):
    """A policy supplied as a file path behaves like the inline form."""
    policy = tmp_path / "read-only.cedar"
    policy.write_text(_READ_ONLY_POLICY)
    shell = strands_shell.Shell(policy_file=str(policy))
    assert shell.run("ls /").status == 0
    assert shell.run("echo hi > /home/lash/x.txt").status != 0


def test_no_policy_is_unchanged():
    """With no policy, writes are allowed (default-allow)."""
    shell = strands_shell.Shell()
    assert shell.run("echo hi > /home/lash/x.txt && cat /home/lash/x.txt").status == 0


def test_policy_and_policy_file_are_mutually_exclusive(tmp_path):
    policy = tmp_path / "p.cedar"
    policy.write_text(_READ_ONLY_POLICY)
    with pytest.raises(ValueError, match="at most one"):
        strands_shell.Shell(policy=_READ_ONLY_POLICY, policy_file=str(policy))


def test_malformed_policy_raises():
    """A malformed policy fails at construction (build())."""
    with pytest.raises(Exception):
        strands_shell.Shell(policy="permit(garbage")
